### PR TITLE
SNOW-869906: Tag package auto-upload feature as experimental, UNIX-only and accessible only via a flag

### PR DESCRIPTION
**Pull request by sfc-gh-vnayak**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: tests/integ/test_packaging.py [hunk 1]
```diff
@@ -497,33 +534,6 @@ def test_add_requirements_with_ranged_requirements_in_yaml(session, ranged_yaml_
         session.add_requirements(ranged_yaml_file)
```
Comment (sfc-gh-vnayak) [inline tests/integ/test_packaging.py:536]: Packages can no longer be added directly during sproc/udf registration because of the new parameter.

File: src/snowflake/snowpark/session.py [hunk 2]
```diff
@@ -716,6 +717,7 @@ def get_packages(self) -> Dict[str, str]:
     def add_packages(
         self,
         *packages: Union[str, ModuleType, Iterable[Union[str, ModuleType]]],
+        upload_custom_packages: bool = False,
         force_push: bool = True,
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/session.py:721]:
```suggestion
        *packages: Union[str, ModuleType, Iterable[Union[str, ModuleType]]],
        *,
        upload_custom_packages: bool = False,
        force_push: bool = True,
```

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/session.py:721]: I would also think that using how we enable or disable sql simplifier should also be how to enable or disable this feature. @sfc-gh-yixie what do you think?

File: tests/integ/test_packaging.py (hunk 1 above)
Comment (sfc-gh-aalam) [reply tests/integ/test_packaging.py:536]: Using how we enable/disable sql simplifier might be useful again here. Make this part of session which is accessible during udf/sproc creation.

File: src/snowflake/snowpark/session.py (hunk 2 above)
Comment (sfc-gh-vnayak) [reply src/snowflake/snowpark/session.py:721]: > 
The extra star is not needed here because packages argument is specified as `*packages`.
For example, `session.add_packages("sktime", True, True)` would result in errors about "True" not being a valid package.

File: src/snowflake/snowpark/session.py [hunk 3]
```diff
@@ -725,8 +727,10 @@ def add_packages(
         :class:`~snowflake.snowpark.udf.UDFRegistration`. See details of
         `third-party Python packages in Snowflake <https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-packages.html>`_.
 
-        Pure Python packages that are not available in Snowflake will be pip installed locally and made available as an
-        import (via zip file on a remote stage).
+        EXPERIMENTAL: Upon setting `upload_custom_packages` to True, Pure Python packages that are not available in
+        Snowflake will be pip installed locally and made available as an import (via zip file on a remote stage).
+        Note that if you wish to use a specific version of pip, you can set the environment variable `PIP_PATH` to your
+        pip executable. This feature is experimental and works well on UNIX systems only, do not use it in production!
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/session.py:733]: I wonder if we need to add a check for non-unix system (if possible at all).

File: src/snowflake/snowpark/session.py [hunk 4]
```diff
@@ -958,6 +980,13 @@ def _resolve_packages(
                             "by ORGADMIN to use Anaconda 3rd party packages. Please follow the instructions at "
                             "https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-packages.html#using-third-party-packages-from-anaconda."
                         )
+                    if not upload_custom_packages:
+                        raise RuntimeError(
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/session.py:984]: Should we highlight this error message change in the release note?

File: src/snowflake/snowpark/session.py (hunk 2 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/session.py:721]: I think that's the intention of adding `*`. We want users to specify packages freely but if they want a force upload they need to do `session.add_packages(package1, package2, force_push=True)`

File: src/snowflake/snowpark/session.py (hunk 3 above)
Comment (sfc-gh-vnayak) [reply src/snowflake/snowpark/session.py:734]: Added this as a warning -> https://github.com/snowflakedb/snowpark-python/pull/959/files#diff-9a82bd4a068fbfc98475b894a9901b9994eb8ca38b2cfda8f2b0427bc368749cR1033-R1036

File: src/snowflake/snowpark/session.py (hunk 2 above)
Comment (sfc-gh-vnayak) [reply src/snowflake/snowpark/session.py:722]: Quick clarification; `*packages,` and `*,` are equivalent. Should I add a test case to ensure that experimental parameters cannot be specified accidentally?

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/session.py:722]: they are not equivalent but it seems they can't be added together: https://peps.python.org/pep-3102/

Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/session.py:722]: @sfc-gh-aalam That sounds a good way to set the parameter on the session because it will impact multiple functions (session.add_requirements, session.udf.register, etc) with a single config. 
Sophie added `session.conf` to add configs at the client. We can use it. 
@sfc-gh-vnayak Then I guess you also need to move `force_push` to the session level?

Comment (sfc-gh-vnayak) [reply src/snowflake/snowpark/session.py]: Switched to parameter.

`force_push` felt too specific to move to session level, instead I have limited usage of `force_push` via `Session` functions only (like `add_requirements` or `add_packages`) and set it to default to `False`.

File: src/snowflake/snowpark/_internal/udf_utils.py [hunk 5]
```diff
@@ -779,7 +779,6 @@ def resolve_imports_and_packages(
     statement_params: Optional[Dict[str, str]] = None,
     source_code_display: bool = False,
     skip_upload_on_content_match: bool = False,
-    force_push: bool = True,
```
Comment (sfc-gh-vnayak) [inline src/snowflake/snowpark/_internal/udf_utils.py:782]: Disallow force_push usage here, only allow it via Session functions.

File: src/snowflake/snowpark/session.py [hunk 6]
```diff
@@ -452,6 +458,14 @@ def sql_simplifier_enabled(self, value: bool) -> None:
             pass
         self._sql_simplifier_enabled = value
 
+    @custom_packages_upload_enabled.setter
+    def custom_packages_upload_enabled(self, value: bool) -> None:
```
Comment (sfc-gh-vnayak) [inline src/snowflake/snowpark/session.py:462]: Below warning is a substitute for @experimental

File: src/snowflake/snowpark/session.py (hunk 4 above)
Comment (sfc-gh-vnayak) [reply src/snowflake/snowpark/session.py:993]: Sure, which changelog section do I add it under?

File: src/snowflake/snowpark/session.py [hunk 7]
```diff
@@ -823,19 +839,28 @@ def clear_packages(self) -> None:
...
+
+        EXPERIMENTAL: Upon setting `Session.custom_packages_upload_enabled` to True, Pure Python packages that are not available in
+        Snowflake will be pip installed locally and made available as an import (via zip file on a remote stage).
+        Note that if you wish to use a specific version of pip, you can set the environment variable `PIP_PATH` to your
+        pip executable. This feature is experimental and works well on UNIX systems only, do not use it in production!
 
         Args:
             file_path: The path of a local requirement file.
-            force_push: Force upload Python packages with native dependencies.
+            force_push: Force upload Python packages with native dependencies (experimental).
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/session.py:864]: nit: can you use the same description for `force_push` for all doc strings:

`force_push: Force upload unavailable Python packages which contain native C/C++ code (experimental).`

File: tests/integ/test_packaging.py [hunk 8]
```diff
@@ -524,6 +563,31 @@ def check_if_package_works() -> str:
...
+    """
+    Assert that unsupported packages can directly be added while registering UDFs.
+    """
+    session.custom_packages_upload_enabled = True
+    with patch.object(session, "_is_anaconda_terms_acknowledged", lambda: True):
+        packages = ["scikit-fuzzy==0.4.2", "snowflake-snowpark-python"]
+        sproc_name = Utils.random_name_for_temp_object(TempObjectType.FUNCTION)
+
+        @sproc(name=sproc_name, packages=packages, return_type=StringType())
+        def check_if_package_works(session_):
```
Comment (sfc-gh-aalam) [inline tests/integ/test_packaging.py:580]: these additions are out of scope for this PR

Comment (sfc-gh-vnayak) [reply tests/integ/test_packaging.py:580]: This test is for ensuring that custom_packages_upload_enabled works for sproc registration; will add it in a later PR

File: tests/integ/test_packaging.py [hunk 9]
```diff
@@ -309,17 +343,19 @@ def test_add_unsupported_requirements_twice_should_not_fail_for_same_requirement
...
         session.add_packages(["scikit-learn==1.2.0"])
         with pytest.raises(ValueError, match="Cannot add dependency package"):
             session.add_packages("sktime==0.20.0")
 
 
 @pytest.mark.skipif(
-    IS_IN_STORED_PROC,
-    reason="Subprocess calls are not allowed within stored procedures",
+    IS_IN_STORED_PROC or IS_WINDOWS,
+    reason="Subprocess calls are not allowed within stored procedures. Unsupported package upload does not work well on Windows.",
```
Comment (sfc-gh-vnayak) [inline tests/integ/test_packaging.py:355]: For all tests which involve unsupported packages, skip if on Windows.

File: src/snowflake/snowpark/session.py [hunk 10]
```diff
@@ -722,7 +736,7 @@ def get_packages(self) -> Dict[str, str]:
     def add_packages(
         self,
         *packages: Union[str, ModuleType, Iterable[Union[str, ModuleType]]],
-        force_push: bool = True,
+        force_push: bool = False,
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/session.py:739]: I like that idea too.

One reason it could be not ideal is when users want to force_push for certain packages but not for others - which does not seem likely to me.
On the other hand, having it a parameter with session object let's us set one behavior for all package upload activities - this sounds good to me.

File: CHANGELOG.md [hunk 11]
```diff
@@ -11,6 +11,7 @@
   - `array_max`
   - `explode_outer`
 - Added support for pure Python packages specified via `Session.add_requirements` or `Session.add_packages`, now usable in stored procedures and UDFs even if packages are not present on the Snowflake Anaconda channel.
+- Added Session parameter `custom_packages_upload_enabled` to enable the feature mentioned above (defaults to `False`).
```
Comment (sfc-gh-aalam) [inline CHANGELOG.md:14]:
```suggestion
- Added support for pure Python packages specified via `Session.add_requirements` or `Session.add_packages`, now usable in stored procedures and UDFs even if packages are not present on the Snowflake Anaconda channel.
  - Added Session parameter `custom_packages_upload_enabled` to enable the feature mentioned above (defaults to `False`).
```
How about moving this in a sub bullet point?

Comment (sfc-gh-vnayak) [reply CHANGELOG.md]: Thanks, changed this. Not sure where to document info about the parameters; where would I add that?

**Review by sfc-gh-aalam — APPROVED** [review]
LGTM

**Review by sfc-gh-yixie — APPROVED** [review]
LGTM.

Circled @sfc-gh-mabrennan to review docstrings.

File: src/snowflake/snowpark/session.py [hunk 12]
```diff
@@ -452,6 +464,16 @@ def sql_simplifier_enabled(self, value: bool) -> None:
             pass
         self._sql_simplifier_enabled = value
 
+    @custom_packages_upload_enabled.setter
+    @experimental_parameter(version="1.6.0")
+    def custom_packages_upload_enabled(self, value: bool) -> None:
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/session.py:469]: How about adding docstring here too to explain what this is for?

File: src/snowflake/snowpark/session.py [hunk 13]
```diff
@@ -452,6 +464,16 @@ def sql_simplifier_enabled(self, value: bool) -> None:
...
+        self._custom_packages_upload_enabled = value
+
+    @custom_packages_force_upload_enabled.setter
+    @experimental_parameter(version="1.6.0")
+    def custom_packages_force_upload_enabled(self, value: bool) -> None:
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/session.py:474]: Same here.

**Comment by sfc-gh-aalam** [comment]
> LGTM.
> 
> Circled @sfc-gh-mabrennan to review docstrings.

I added @sfc-gh-straut because Mary is out.

File: src/snowflake/snowpark/session.py (hunk 12 above)
Comment (sfc-gh-vnayak) [reply src/snowflake/snowpark/session.py:469]: Sure, added

File: src/snowflake/snowpark/session.py [hunk 14]
```diff
@@ -951,6 +979,13 @@ def _resolve_packages(
                             "by ORGADMIN to use Anaconda 3rd party packages. Please follow the instructions at "
                             "https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-packages.html#using-third-party-packages-from-anaconda."
                         )
+                    if not self.custom_packages_upload_enabled:
+                        raise RuntimeError(
+                            f"Cannot add package {package_req} because it is not available in Snowflake "
+                            f"and Session parameter 'custom_packages_upload_enabled' is set to False. To upload these packages, you can "
+                            f"set it to True or find the directory of these packages and add it via session.add_import(). See details at "
```
Comment (sfc-gh-straut) [inline src/snowflake/snowpark/session.py:986]: Should True and the method name be formatted for monospace? Also, I'd use the following for the method name

> ``session.add_import`` method.

File: src/snowflake/snowpark/session.py [hunk 15]
```diff
@@ -1092,9 +1128,12 @@ def _upload_unsupported_packages(
...
             )
 
-            if len(native_packages) > 0 and not force_push:
+            if (
+                len(native_packages) > 0
+                and not self._custom_packages_force_upload_enabled
+            ):
                 raise ValueError(
-                    "Your code depends on native dependencies, it may not work on Snowflake! Use option `force_push` "
+                    "Your code depends on native dependencies, it may not work on Snowflake! Set Session parameter 'custom_packages_force_upload_enabled' to True "
```
Comment (sfc-gh-straut) [inline src/snowflake/snowpark/session.py:1136]: "True" should probably be monospace (code) font here.

File: tests/integ/test_packaging.py [hunk 16]
```diff
@@ -326,12 +329,44 @@ def test_add_requirements_twice_should_fail_if_packages_are_different(
...
+        with pytest.raises(
+            RuntimeError,
+            match="parameter 'custom_packages_upload_enabled' is set to False",
+        ):
+            session.add_requirements(test_files.test_unsupported_requirements_file)
+
+
+@pytest.mark.skipif(
+    IS_IN_STORED_PROC or IS_WINDOWS,
+    reason="Subprocess calls are not allowed within stored procedures. Custom package upload does not work well on Windows.",
```
Comment (sfc-gh-straut) [inline tests/integ/test_packaging.py:349]: Is there anything more specific to say (here and elsewhere) about why this doesn't work well on Windows?

**Review by sfc-gh-straut — COMMENTED** [review]
It looks like you're setting monospace font using both single- and double-backticks. Unless that's needed based on context, I'd use one style only.

File: src/snowflake/snowpark/session.py (hunk 12 above)
Comment (sfc-gh-vnayak) [reply src/snowflake/snowpark/session.py:474]: Docstrings are moved to the function which ends up being rendered (the setter)

File: tests/integ/test_packaging.py (hunk 16 above)
Comment (sfc-gh-vnayak) [reply tests/integ/test_packaging.py:349]: This is being investigated (https://github.com/snowflakedb/snowpark-python/pull/963), a concrete root cause has not been found yet.

File: src/snowflake/snowpark/session.py (hunk 15 above)
Comment (sfc-gh-vnayak) [reply src/snowflake/snowpark/session.py]: I see that errors aren't usually monospaced in the repo as code errors are unlikely to be rendered in code font.

File: src/snowflake/snowpark/session.py [hunk 17]
```diff
@@ -823,19 +847,22 @@ def clear_packages(self) -> None:
...
         Adds a `requirement file <https://pip.pypa.io/en/stable/user_guide/#requirements-files>`_
         that contains a list of packages as dependencies of a user-defined function (UDF). Pure Python packages that
         are not available in Snowflake will be pip installed locally and made available as an import (via zip file
         on a remote stage).
 
-         Note that this function also supports addition of requirements via a `conda environment yaml file
-         <https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#create-env-file-manually>_`.
+        Note that this function also supports addition of requirements via a `conda environment yaml file <https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#create-env-file-manually>_`.
+
+        EXPERIMENTAL: Upon setting `Session.custom_packages_upload_enabled` to True, Pure Python packages that are not available in
```
[the remark this answers is not in the document]
Comment (sfc-gh-vnayak) [reply src/snowflake/snowpark/session.py]: Thanks, fixed this

File: src/snowflake/snowpark/session.py [hunk 18]
```diff
@@ -731,8 +755,10 @@ def add_packages(
         :class:`~snowflake.snowpark.udf.UDFRegistration`. See details of
         `third-party Python packages in Snowflake <https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-packages.html>`_.
 
-        Pure Python packages that are not available in Snowflake will be pip installed locally and made available as an
-        import (via zip file on a remote stage).
+        EXPERIMENTAL: Upon setting `Session.custom_packages_upload_enabled` to True, Pure Python packages that are not available in
```
[the remark this answers is not in the document]
Comment (sfc-gh-vnayak) [reply src/snowflake/snowpark/session.py]: Fixed this to be formatted as the others are

File: src/snowflake/snowpark/session.py [hunk 19]
```diff
@@ -452,6 +464,20 @@ def sql_simplifier_enabled(self, value: bool) -> None:
             pass
         self._sql_simplifier_enabled = value
 
+    @custom_packages_upload_enabled.setter
+    @experimental_parameter(version="1.6.0")
+    def custom_packages_upload_enabled(self, value: bool) -> None:
+        """Set to ``True`` to automatically upload custom packages.
+        If set to ``True``, any package mentioned during UDF/Stored Procedure registration or via Session.add_packages() or Session.add_requirements() will be pip installed, zipped and made available as an import via a remote stage."""
```
[the remark this answers is not in the document]
Comment (sfc-gh-vnayak) [reply src/snowflake/snowpark/session.py]: Pip is not zipping the files; I updated this sentence so that it is less ambiguous.

File: src/snowflake/snowpark/session.py [hunk 20]
```diff
@@ -1120,7 +1160,7 @@ def _upload_unsupported_packages(
                 self._tmpdir_handler.cleanup()
             raise RuntimeError(
                 f"Unable to auto-upload packages: {packages}, Error: {e} | NOTE: Alternatively, you can find the "
-                f"directory of these packages and add it via session.add_import(). See details at "
+                f"directory of these packages and add it via Session.add_import. See details at "
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/session.py:1163]: `session.add_import` here is formatted differently from the in on line 990. can you make them consistent
